packages/datadog-plugin-graphql/src/index.js

```diff
@@ -34,8 +34,13 @@ function createWrapExecute (tracer, config, defaultFieldResolver) {
       contextValue._datadog_graphql = { source, span, fields: {} }
 
       return call(execute, span, this, [args], (err, res) => {
-        finishResolvers(contextValue)
-        finish(err || (res && res.errors && res.errors[0]), span)
+        finishResolvers(contextValue, config)
+
+        finish(err || (res && res.errors && res.errors[0]), span, undefined, (span) => config.hooks.execute(
```

Comment:
Can this be rewritten in a way that doesn't require passing `undefined` arguments? I think it could be called right before calling `finish`.

Reply:
I'm passing the hook down to `finish` this way for one reason. I think the hook should be called after the plugin adds any possible error tags to the span but before the `finish` is called on the span.
```js
function finish (error, span, finishTime, hook) {
  if (error) {
    span.addTags({
      'error.type': error.name,
      'error.msg': error.message,
      'error.stack': error.stack
    })
  }

  // Call hook before span#finish is called but after errors have possibly been added
  if (hook) hook(span)

  span.finish(finishTime)
}
```

This would possibly enable the client to remove the error tag if necessary. There's no API currently for removing tags but I think it might be possible to go around it with:
```js
span.addTags({
  'error.type': undefined,
  'error.msg': undefined,
  'error.stack': undefined
})
```

The use case for this would be, as discussed in the issue, to determine whether a graphql operation failed due to client or server errors. Client errors should not result in the execution span to be marked as an error.

Reply:
We could get rid of these `undefined` values if we changed the order of params:
```js
function finish (error, span, finishTime, hook) { ... } // old
function finish (error, span, hook, finishTime) { ... } // new
```
However, undefined values might appear in other places then (for which there's no hook defined, yet).

Reply:
That makes sense, I didn't think about the error. Maybe then it would be better to split `finish` and add the error separately? Then it could look like this:

```js
addError(err)
executeHook(span, context, res)
span.finish()
```

Although I'm not sure the best way to prevent the error is to set every field to `undefined` as it feels a bit hacky. I'll take a look to see if it could be done otherwise with the current API.

Reply:
Ok I think the best syntax for now would probably be to rewrite how the error is added to use `span.setTag('error', error)` instead. This is a new syntax that was not supported when the plugin was written. Then, to remove the error all that will be needed is to do `span.setTag('error', false)` or `span.setTag('error', 0)`.

Reply:
> Maybe then it would be better to split finish and add the error separately? 

Should it be split for all wrappers that use `finish` or just in the case of the execution wrapper?

Reply:
Replaced `addTags` with `setTag('error', error)`: https://github.com/DataDog/dd-trace-js/pull/909/commits/f6c57e27066b00732f50d2c8cd82819c61ecc689.

Reply:
As suggested in https://github.com/DataDog/dd-trace-js/issues/884#issuecomment-612568623, I think it's worth considering what the default behaviour should be and how to supply it.

Proposal: move the `setTag('error', error)` into the default (currently noop) `execute` callback. Consequences:
* consumers can opt-out of the current behaviour by supplying their own callback
* consumers don't need to un-do the `setTag` call at runtime
* the default `execute` callback can serve as documentation for how to use the hook

Reply:
> Should it be split for all wrappers that use finish or just in the case of the execution wrapper?

I would say for now whatever results in the least amount of code. When other hooks are implemented, part of this will probably be refactored anyway so for now let's keep it simple.

> consumers can opt-out of the current behaviour by supplying their own callback

In general the hooks are general-purpose functions that are meant to augment the spans from the plugins, but not the auto-instrumentation. In that sense I think a config option would be more adequate for this specific use case.

Reply:
Addressed in https://github.com/DataDog/dd-trace-js/pull/909/commits/9af35067071f369cb00e964903fb1f29926d40b3.